### DISCUSSION TITLE
Creating custom JDBC wrapper for serialization/deserialization

### DISCUSSION THREAD

**Discussion by pranjalc92** [vladimir-bukhtoyarov/bucket4j#505] [Q&A] [answered]
We have a database/data source that is not supported by bucket4j. 
Thus we would like to create the bucket using data stored in the database and update it in the database after every request. This will happen every time a REST API request is received. The requirement to store in a database is primarily because we have multiple machines in a cluster and would like to define rate limits at the cluster level and not the individual machine level. 

1. Is it fine to use `LocalBucket.fromJsonCompatibleSnapshot` to deserialize the data from the DB? 
2. Are we losing any features from Bucket4J while doing so?
3. How would bucket refill work in this case? My understanding is that this will be primarily done when `tryConsume` is called after creating the bucket from the state stored in the DB.

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @pranjalc92

> Is it fine to use LocalBucket.fromJsonCompatibleSnapshot to deserialize the data from the DB?

I suppose that it will be unpleasure excersize. The main problem will be correctness in concurrent environment.

> Are we losing any features from Bucket4J while doing so?

At least  [on-the-fly configuration replacement](https://bucket4j.com/8.13.1/toc.html#configuration-replacement) and [implicit configuration replacement](https://bucket4j.com/8.13.1/toc.html#implicit-configuration-replacement). And you will not be able to reuse AbstractDistributedBucketTest to check whether you solution is correct or not. So I highly reccomend to extend one of base classes(choose particular depending from concurrency features that your database provides) instead of writing all code from scratch:
* AbstractCompareAndSwapBasedProxyManager
* AbstractSelectForUpdateBasedProxyManager
* AbstractLockBasedProxyManager
* AbstractProxyManager

Extending one of base class will allow to resuse the TCK and reuse ideas from already implemented integrations.

> How would bucket refill work in this case?

Refill should works fine, in way way as you proposed.
